allennlp/modules/text_field_embedders/basic_text_field_embedder.py

```diff
@@ -36,12 +36,13 @@ def get_output_dim(self) -> int:
         return output_dim
 
     def forward(self, text_field_input: Dict[str, torch.Tensor], num_wrapping_dims: int = 0) -> torch.Tensor:
-        if self._token_embedders.keys() != text_field_input.keys():
-            message = "Mismatched token keys: %s and %s" % (str(self._token_embedders.keys()),
-                                                            str(text_field_input.keys()))
+        if not set(self._token_embedders.keys()).issubset(set(text_field_input.keys())):
```

Comment:
Does it not work without this change?  I'd prefer matching the set exactly, mainly to help avoid user error in writing the configuration file.  But, if others also think this is a good change, or if it's necessary to use ELMo in advanced ways, we can change it.

Reply:
so the ELMo object requires the "elmo" key in the token indexer, and the previous behavior mandated that you must also have an "elmo" key in the embedder as well.

However, you might not want to use the input ELMo embeddings and use only ELMo embeddings at the 2nd LSTM's output --- this lets you do that.

I agree that this isn't ideal from an API perspective, but unsure if there's another good solution? Would be all ears to suggestions.

Reply:
A better way - include the `token_indexer` for `elmo_characters`, but pop it off the dictionary in `forward`. Then have a flag like `use_elmo` which dictates whether you actually use it or not, passing it directly to an `Elmo()` instance with `num_output_representations=2`. This way of using elmo doesn't really fit the API, but I agree with Matt that it would be better not to dilute how `textfield_embedders` are used, because they are already kind of confusing for new users.

Reply:
@DeNeutoy : i like your way better as well, thanks!

It might be a bit confusing to the user if they include elmo in their embedder and then see that it doesn't get used, but I think this should be ok with proper documentation / raising an informative error message if elmo is used in the embedder...